zerver/lib/user_groups.py

```diff
@@ -173,10 +155,14 @@ def access_user_group_for_update(
         raise JsonableError(_("Insufficient permission"))
 
     assert permission_setting in NamedUserGroup.GROUP_PERMISSION_SETTINGS
-    if permission_setting in [
-        "can_manage_group",
-        "can_add_members_group",
-    ] and check_permission_for_managing_all_groups(user_group, user_profile):
+    if (
+        permission_setting
+        in [
+            "can_manage_group",
+            "can_add_members_group",
+        ]
+        and user_profile.can_manage_all_groups()
+    ):
```

Comment:
This seems logically correct, but I think a lot cleaner way to express this check would be for `user_has_permission_for_group_setting` to be responsible for checking if the permission is `can_manage_group`, and if so, checking `user_profile.can_manage_all_groups()`, and then the can-manage-all-groups behavior here would be implemented through the block 10 lines down with `if not user_has_permission and permission_setting == "can_add_members_group":`.

That just feels like the abstractions would be a lot more natural.

Reply:
If I'm understanding the comment correctly, we want to move the can_manage_all_groups check to `user_has_permission_for_group_setting`. I have similar comments to https://github.com/zulip/zulip/pull/31928#discussion_r1796887287 here also. It would nice to just keep `user_has_permission_for_group_setting` for checking the value directly.